src/utils/DecoyDatabase.cpp

```diff
@@ -141,45 +149,93 @@ class TOPPDecoyDatabase :
       //-------------------------------------------------------------
       // calculations
       //-------------------------------------------------------------
-      while (f.readNext(protein))
+      while (f.readNext(entry))
       {
-        if (identifiers.find(protein.identifier) != identifiers.end())
+        if (identifiers.find(entry.identifier) != identifiers.end())
         {
-          LOG_WARN << "DecoyDatabase: Warning, identifier '" << protein.identifier << "' occurs more than once!" << endl;
+          LOG_WARN << "DecoyDatabase: Warning, identifier '" << entry.identifier << "' occurs more than once!" << endl;
         }
-        identifiers.insert(protein.identifier);
+        identifiers.insert(entry.identifier);
 
         if (append)
         {
-          f.writeNext(protein);
+          f.writeNext(entry);
         }
       
         // identifier
-        protein.identifier = getIdentifier_(protein.identifier, decoy_string, decoy_string_position_prefix);
-      
+        entry.identifier = getIdentifier_(entry.identifier, decoy_string, decoy_string_position_prefix);
+
         // sequence
-        if (shuffle)
+        if (input_type == seq_type::RNA)
         {
-          String temp;
-          Size x = protein.sequence.size();
-          srand(time(0));
-          while (x != 0)
+          string quick_seq = entry.sequence;
+          bool five_p = (entry.sequence.front() == 'p');
+          bool three_p = (entry.sequence.back() == 'p');
+          if (five_p) //we don't want to reverse terminal phosphates
+          {
+            quick_seq.erase(0,1);
+          }
+          if (three_p)
+          {
+            quick_seq.pop_back();
+          }
+          vector<String> tokenized;
+          smatch m;
+          while (regex_search(quick_seq, m, std::regex("[^\\[]|(\\[[^\\[\\]]*\\])"))){
+            tokenized.push_back(m.str(0));
+            quick_seq = m.suffix();
+          }
+
+          if (shuffle)
           {
-            Size y = rand() % x;
-            temp += protein.sequence[y];
-            --x;
-            protein.sequence[y] = protein.sequence[x]; // overwrite consumed position with last position (about to go out of scope for next dice roll)
+            random_shuffle(tokenized.begin(),tokenized.end());
           }
+          else  // reverse
+          {
+            reverse (tokenized.begin(), tokenized.end()); //reverse the tokens
+            if (five_p)  //add back 5'
```

Comment:
Add "}" before this, so 5'/3' phosphates are added back for shuffled and reversed sequences.